### ISSUE TITLE
concern about jquery memory leaks

### ISSUE THREAD

**Issue by patleb** [turbolinks/turbolinks#144]
It's more a concern than a current issue, but I was wondering if anyone ever encountered a memory leak with Turbolinks and jQuery since it replaces the content without going through jQuery `.html()` method? As per this [post](https://makandracards.com/makandra/31325-how-to-create-memory-leaks-in-jquery), this is what I would expect if I was using Turbolinks and navigating from page to page without any refresh.

**Comment by nateberkopec** [comment]
The article you linked is very much specific to JQuery - since Turbolinks does not create new JQuery objects, it does not add anything to `$.cache`, and does not cause a memory leak in the way that article describes.

**Comment by patleb** [comment]
No, it's specific to Turbolinks with jQuery, not Turbolinks only. I understand this isn't specific to this library, but seamless integration with jQuery without memory leaks surprises seems to me like a good enough requirement for a library which is more often than not used alongside jQuery in the Rails projects.

**Comment by nateberkopec** [comment]
The memory leak described in that article is caused when manipulating JQuery objects. Turbolinks does not create or manipulate JQuery objects *even when JQuery is present*. Therefore, it cannot cause this particular memory leak.

**Comment by nateberkopec** [comment]
Aha! Now I get it.

```javascript
// you append a JQuery object to the DOM somewhere in your application
var $element = $('<div></div>');
$element.on('click', function() { ... });
$element.appendTo(document.body);
// later, in turbolinks:
document.body.innerHTML = newHtml;
```

AFAICT you *must* append a Jquery object *with an event handler*  to the DOM for this to cause a leak. Simply using JQuery and Turbolinks in the same page won't cause a problem (this is what I thought you were originally proposing).

However, I cannot reproduce this memory leak in JQuery 2.x on latest Chrome. [Here's a JSFiddle](https://jsfiddle.net/2b915t9k/). Run it, and look at the heap in Chrome Timeline. It looks like this:

<img width="1272" alt="screen shot 2016-10-08 at 10 59 59 am" src="https://cloud.githubusercontent.com/assets/845662/19214680/674992f6-8d46-11e6-965c-18e62b16e1fd.png">

Doesn't look leaky to me - we can clearly see some GC happening. [I get the same result with innerHTML](https://jsfiddle.net/ovm7kux6/). 

Now, you shouldn't be using `on` this way with Turbolinks anyway. [We already say in the README:](https://github.com/turbolinks/turbolinks#running-javascript-when-a-page-loads)

> When possible, avoid using the turbolinks:load event to add event listeners directly to elements on the page body. Instead, consider using event delegation to register event listeners once on document or window.

If you do what the README says, you should never encounter this memory leak.

However, maybe some people in the wild are doing this:

```javascript
document.addEventListener("turbolinks:load", function() {
  var $element = $('<div></div>');
  $element.on('click', function() { ... });
  $element.appendTo(document.body);
})
```

So, in summary:

* Turbolinks doesn't recommend adding event listeners this way anyway.
* I can't repro the leak on JQuery 2.x, though I'm not really experienced on this kind of thing so maybe I just messed it up somehow.
* Turbolinks won't leak if you're using JQuery with it in the way the README recommends.

**Comment by nateberkopec** [comment]
[And here's a really long discussion about this issue on turbolinks-classic.](https://github.com/turbolinks/turbolinks-classic/pull/502)

**Comment by nateberkopec** [comment]
https://github.com/jquery/jquery/pull/1565 

I can repro this leak only with JQuery <= 2.1. [Here's a gist, beware this is like a 1GB leaker](https://gist.github.com/nateberkopec/bbbbd12fe90f8e5d03728dcf7861347b).

**Comment by patleb** [comment]
Wow, that's a very thorough answer. I wasn't expecting that much, since I was pointing this as a concern. Which isn't a concern anymore, thx a lot!

**Comment by nateberkopec** [comment]
@patleb I love this kinda troubleshooting

Still, I don't think TL needs to do anything about it. It only leaks *if* you're doing something TL says you shouldn't do *and* using a 2 year old+ JQuery version.

**Comment by Ben-Abbes** [comment]
@nateberkopec I am using jquery 3 version. Do I always have to worry about this memory leak? Or does it only concern versions prior to jquery 2?
